utils/regression-analysis/src/main.rs

```diff
@@ -0,0 +1,221 @@
+
+use std::{
+    collections::BTreeMap,
+    fs,
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+
+use clap::{Parser, Subcommand};
+
+use quick_xml::de::from_str;
+
+mod junit_tree;
+mod output;
+
+use common::TestSuites;
+use tabled::{Style, Table};
+
+const PALLET_NAMES: [&str; 7] = [
+    "pallet-gas",
+    "pallet-gear",
+    "pallet-gear-debug",
+    "pallet-gear-messenger",
+    "pallet-gear-program",
+    "pallet-gear-payment",
+    "pallet-usage",
+];
+
+const PREALLOCATE: usize = 1_000;
+
+const TEST_SUITES_TEXT: &str = "Test suites";
+
+#[derive(Parser)]
+struct Cli {
+    #[clap(subcommand)]
+    command: Commands,
+}
+
+#[derive(Subcommand)]
+enum Commands {
+    CollectData {
+        #[clap(long)]
+        data_folder_path: PathBuf,
+        #[clap(long)]
+        output_path: PathBuf,
+        #[clap(long)]
+        disable_filter: bool,
+    },
+    Compare {
+        #[clap(long)]
+        data_path: PathBuf,
+        #[clap(long)]
+        current_junit_path: PathBuf,
+        #[clap(long)]
+        disable_filter: bool,
+    },
+}
+
+fn build_tree<P: AsRef<Path>>(
+    disable_filter: bool,
+    path: P,
+) -> BTreeMap<String, BTreeMap<String, f64>> {
+    let filter = |pallet_name: &str| {
+        if disable_filter {
+            return true;
+        }
+
+        PALLET_NAMES.iter().any(|&name| name == pallet_name)
+    };
+
+    let junit_xml = std::fs::read_to_string(path).unwrap();
+    let test_suites: TestSuites = from_str(&junit_xml).unwrap();
+    let total_time: BTreeMap<_, _> = [(
+        String::from("Total time"),
+        f64::from_str(&test_suites.time).unwrap(),
+    )]
+    .into();
+    let mut result = junit_tree::build_tree(filter, test_suites);
+    result.insert(String::from(TEST_SUITES_TEXT), total_time);
+    result
+}
+
+fn median(values: &[u64]) -> u64 {
+    assert!(!values.is_empty());
+
+    let len = values.len();
+    if len % 2 == 0 {
+        let i = len / 2;
+        values[i - 1] / 2 + values[i] / 2 + values[i - 1] % 2 + values[i] % 2
+    } else {
+        values[len / 2]
+    }
+}
+
+fn collect_data<P: AsRef<Path>>(
+    data_folder_path: P,
+    output_path: P,
+    disable_filter: bool,
+    preallocate: usize,
+) {
+    let mut statistics: BTreeMap<String, BTreeMap<String, Vec<u64>>> = BTreeMap::default();
+    for entry in fs::read_dir(data_folder_path).unwrap() {
+        let executions = build_tree(disable_filter, &entry.unwrap().path());
+        executions.iter().for_each(|(key, times)| {
+            if !statistics.contains_key(key) {
+                statistics.insert(key.clone(), Default::default());
+            }
+
+            let previous_times = statistics.get_mut(key).unwrap();
+            times.iter().for_each(|(key, &time)| {
+                let time = (1_000_000_000.0 * time) as u64;
+
+                if let Some(time_vec) = previous_times.get_mut(key) {
+                    let i = match time_vec.binary_search(&time) {
```

Comment:
[Result::into_ok_or_err()](https://doc.rust-lang.org/std/result/enum.Result.html#method.into_ok_or_err) is unstable unfortunately 😄

Reply:
yep )